Python/Product/PythonTools/PythonTools/Environments/AddVirtualEnvironmentOperation.cs

```diff
@@ -0,0 +1,183 @@
+            _virtualEnvPath = virtualEnvPath ?? throw new ArgumentNullException(nameof(virtualEnvPath));
+            _baseInterpreter = baseInterpreterId ?? throw new ArgumentNullException(nameof(baseInterpreterId));
+            _useVEnv = useVEnv;
+            _installReqs = installRequirements;
+            _reqsPath = requirementsPath;
+            _registerAsCustomEnv = registerAsCustomEnv;
+            _customEnvName = customEnvName;
+            _setAsCurrent = setAsCurrent;
+            _setAsDefault = setAsDefault;
+            _viewInEnvWindow = viewInEnvWindow;
+            _output = output;
+            _statusCenter = _site.GetService(typeof(SVsTaskStatusCenterService)) as IVsTaskStatusCenterService;
+            _registry = _site.GetComponentModel().GetService<IInterpreterRegistryService>();
+            _options = _site.GetComponentModel().GetService<IInterpreterOptionsService>();
+            _logger = _site.GetService(typeof(IPythonToolsLogger)) as IPythonToolsLogger;
+        }
+
+        public async Task RunAsync() {
+            var outputWindow = OutputWindowRedirector.GetGeneral(_site);
+            var taskHandler = _statusCenter.PreRegister(
+                new TaskHandlerOptions() {
+                    ActionsAfterCompletion = CompletionActions.RetainAndNotifyOnFaulted | CompletionActions.RetainAndNotifyOnRanToCompletion,
+                    Title = Strings.VirtualEnvStatusCenterCreateTitle.FormatUI(PathUtils.GetFileOrDirectoryName(_virtualEnvPath)),
+                    DisplayTaskDetails = (t) => { outputWindow.ShowAndActivate(); }
+                },
+                new TaskProgressData() {
+                    CanBeCanceled = false,
+                    ProgressText = Strings.VirtualEnvStatusCenterCreateProgressPreparing,
+                    PercentComplete = null,
+                }
+            );
+
+            var task = CreateVirtualEnvironmentAsync(taskHandler);
+            taskHandler?.RegisterTask(task);
+            _site.ShowTaskStatusCenter();
+        }
+
+        private async Task CreateVirtualEnvironmentAsync(ITaskHandler taskHandler) {
+            IPythonInterpreterFactory factory = null;
+
+            bool failed = true;
+            var baseInterp = _registry.FindInterpreter(_baseInterpreter);
+
+            try {
+                factory = await VirtualEnv.CreateAndAddFactory(
+                    _site,
+                    _registry,
+                    _options,
+                    _project,
+                    _virtualEnvPath,
+                    baseInterp,
+                    _registerAsCustomEnv,
+                    _customEnvName,
+                    _useVEnv
+                );
+
+                if (factory != null) {
+                    if (_installReqs && File.Exists(_reqsPath)) {
+                        await InstallPackagesAsync(taskHandler, factory);
+                    }
+
+                    await _site.GetUIThread().InvokeTask(async () => {
```

Comment:
In some cases we use `GetUIThread().InvokeTask()`, and in some other it is `Dispatcher.Invoke`.

Reply:
The `DependencyObject` derived classes use dispatcher since it's provided by `DependencyObject`, but the other classes (the operation classes) use `GetUIThread`.

I figured that using the dispatcher was a 'lesser' dependency than getting the ui thread service. Is it better to use `GetUIThread` there too?